### PR TITLE
iOS always on

### PR DESCRIPTION
**Pull request by diegoreymendez**
Task/Issue URL: https://app.asana.com/0/1203137811378537/1204886111391108/f

## Description

Implements on-demand through always-on.

This feature is disabled by default and needs to be enabled in Network Protection's debug settings menu.

## Always on is OFF by default

1. Install the app fresh.
2. Go to Network Protection debug settings and ensure always ON is off by default.
1. Start NetP, make sure it connects.
3. Go to System Settings > VPN > Details and make sure always-on is disabled for our connection.

## Our App can disable by just stopping.

1. Make sure always-on is enabled in Network Protection's debug menu.
1. Start NetP, make sure it connects.
2. Go to System Settings > VPN > Details and make sure always-on is enabled for our connection.
3. Stop NetP from our app and make sure it stops fine.

## System Settings need to disable always-on.

1. Make sure always-on is enabled in Network Protection's debug menu.
1. Start NetP
5. Try to stop it from system settings without first disabling always-on, it should reconnect.
6. Go into the details for the VPN connection in system settings, disable always-ON and make sure it disconnects.

## Failure scenario

1. Make sure always-on is enabled in Network Protection's debug menu.
2. Start NetP
3. Make sure it connects.
4. Crash the network extension.
5. Make sure it reconnects (will be started when there's network activity).

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

### CODE REVIEW

File: DuckDuckGo/NetworkProtectionTunnelController.swift [hunk 1]
```diff
@@ -52,13 +52,19 @@ final class NetworkProtectionTunnelController: TunnelController {
     }
 
     func stop() async {
+        guard let tunnelManager = await loadTunnelManager() else {
+            return
+        }
```
Comment (diegoreymendez) [inline DuckDuckGo/NetworkProtectionTunnelController.swift:58]: If there's no tunnel manager, there's no connection to stop.

File: DuckDuckGo/NetworkProtectionTunnelController.swift [hunk 2]
```diff
@@ -52,13 +52,19 @@ final class NetworkProtectionTunnelController: TunnelController {
...
         do {
-            try await ConnectionSessionUtilities.activeSession()?.stopVPNTunnel()
+            try await disableOnDemand(tunnelManager: tunnelManager)
         } catch {
             #if DEBUG
             errorStore.lastErrorMessage = error.localizedDescription
             #endif
         }
+
+        tunnelManager.connection.stopVPNTunnel()
```
Comment (diegoreymendez) [inline DuckDuckGo/NetworkProtectionTunnelController.swift:68]: First disable on-demand, then stop the tunnel.

File: DuckDuckGo/NetworkProtectionTunnelController.swift [hunk 3]
```diff
@@ -177,6 +183,26 @@ final class NetworkProtectionTunnelController: TunnelController {
...
+
+        try await tunnelManager.saveToPreferences()
+    }
+
+    @MainActor
+    func disableOnDemand(tunnelManager: NETunnelProviderManager) async throws {
+        tunnelManager.isOnDemandEnabled = false
+
+        try await tunnelManager.saveToPreferences()
+    }
```
Comment (diegoreymendez) [inline DuckDuckGo/NetworkProtectionTunnelController.swift:208]: On-demand support.

File: DuckDuckGo/NetworkProtectionDebugViewController.swift [hunk 4]
```diff
@@ -28,12 +28,14 @@ import NetworkProtection
 final class NetworkProtectionDebugViewController: UITableViewController {
     private let titles = [
         Sections.keychain: "Keychain",
+        Sections.debugFeature: "Debug Features",
```
Comment (diegoreymendez) [inline DuckDuckGo/NetworkProtectionDebugViewController.swift:35]: The changes in this file are for being able to toggle the feature on.

File: DuckDuckGo/NetworkProtectionDebugViewController.swift [hunk 5]
```diff
@@ -95,19 +104,11 @@ final class NetworkProtectionDebugViewController: UITableViewController {
...
-                cell.textLabel?.text = "Enable NetP > Controller Failure"
-            case .tunnelFailure:
-                cell.textLabel?.text = "Enable NetP > Tunnel Failure"
-            case .crashFatalError:
-                cell.textLabel?.text = "Tunnel: Crash (Fatal Error)"
-            case .crashMemory:
-                cell.textLabel?.text = "Tunnel: Crash (CPU/Memory)"
-            case .none:
-                break
-            }
```
Comment (diegoreymendez) [inline DuckDuckGo/NetworkProtectionDebugViewController.swift:110]: I didn't want to do this change, but I was getting a cyclomatic complexity alert in this method.

File: DuckDuckGo/NetworkProtectionDebugViewController.swift [hunk 6]
```diff
@@ -133,21 +135,71 @@ final class NetworkProtectionDebugViewController: UITableViewController {
...
+            didSelectDebugFeature(at: indexPath)
         case .simulateFailure:
-            switch SimulateFailureRows(rawValue: indexPath.row) {
-            case .controllerFailure: simulateFailure(option: .controllerFailure)
-            case .tunnelFailure: simulateFailure(option: .tunnelFailure)
-            case .crashFatalError: simulateFailure(option: .crashFatalError)
-            case .crashMemory: simulateFailure(option: .crashMemory)
-            case .none: return
-            }
+            didSelectSimulateFailure(at: indexPath)
```
Comment (diegoreymendez) [inline DuckDuckGo/NetworkProtectionDebugViewController.swift:139]: Same as above - I did this to remove the cyclomatic complexity warnings.

File: DuckDuckGo/NetworkProtectionDebugViewController.swift [hunk 7]
```diff
@@ -95,19 +104,11 @@ final class NetworkProtectionDebugViewController: UITableViewController {
...
-            case .tunnelFailure:
-                cell.textLabel?.text = "Enable NetP > Tunnel Failure"
-            case .crashFatalError:
-                cell.textLabel?.text = "Tunnel: Crash (Fatal Error)"
-            case .crashMemory:
-                cell.textLabel?.text = "Tunnel: Crash (CPU/Memory)"
-            case .none:
-                break
-            }
+            configure(cell, forSimulateFailureAtRow: indexPath.row)
```
Comment (diegoreymendez) [inline DuckDuckGo/NetworkProtectionDebugViewController.swift:111]: I didn't want to do this change, but I was getting a cyclomatic complexity alert in this method, due to my additions.

I could've added a flag to disable the warning but I figured we'll probably prefer to do things right.

Comment (graeme) [reply DuckDuckGo/NetworkProtectionDebugViewController.swift:111]: This change is causing release builds to fail. You’ll need to either move the `configure` function above the `NETWORK_PROTECTION` compiler flag. Sorry, this is a bit fiddly. I need to address this properly as the options shouldn’t be visible at all in release builds…

**Review by graeme — APPROVED** [review]
Looks good, thanks for doing this! Once you’ve fixed the release build (probably in the way I suggested), it’s good to go.

File: DuckDuckGo/NetworkProtectionDebugViewController.swift (hunk 7 above)
Comment (diegoreymendez) [reply DuckDuckGo/NetworkProtectionDebugViewController.swift:111]: Thank you, fixed it for release builds so that they show an empty VC for now.

**Pull request merged**